docs: Consolidate and update HIP catch test build documentation (#2198)
ROCm/rocm-systems: projects/hip-tests/README.md

```
`hip-tests` for AMD platform now rely on `amdclang` to build, which is shipped with ROCm installation.
Although individual tests will compile with `hipcc`, ideally you should use `amdclang++`.

2. Build the catch tests
```bash
$ cd "$HIP_TESTS_DIR"
$ mkdir -p build; cd build
$ cmake ../catch -DHIP_PLATFORM=amd -DCMAKE_PREFIX_PATH=<HIP-Installed-Path> -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER=amdclang++ -DCMAKE_C_COMPILER=amdclang -DCMAKE_HIP_COMPILER=amdclang++ -DOFFLOAD_ARCH_STR="--offload-arch=gfxXXXX"
$ make -j32 build_tests
$ ctest # run tests
```

HIP catch tests are built under the folder `$HIP_TESTS_DIR/build`.

### Build HIP Catch2 standalone test

`standalone_main.cc` is now removed, and we use the main shipped with Catch2 to build standalone tests.
The test suite has moved to Catch2v3 (v3.8.1 to be exact).
Moving from v2 to v3 came with some fundamental changes in how Catch2 interacts with hip-tests.
Starting with, it is no longer a single header, it is now a library, which needs to be linked to the test exe.

If you are on your personal machine, it is highly recommended to install Catch2v3 (v3.8.1) locally on your system.

> This helps skip the download/build part in hip-tests and results in faster builds overall.

Reviewer: ```suggestion
`standalone_main.cc`, which was previously used, has been removed. We now rely on the main function provided by Catch2 to build standalone tests. The test suite has been upgraded to Catch2 v3 (specifically v3.8.1). This transition from v2 to v3 introduced several significant changes in how Catch2 integrates with hip-tests. Most importantly, Catch2 is no longer a single-header library; it is now a compiled library that must be linked into the test executable.

For developers working on personal machines, installing Catch2 v3 (v3.8.1) system-wide is strongly recommended. This avoids the download and build steps within hip-tests and leads to noticeably faster build times.

```

HIP Catch2 supports building standalone tests, for example,

### Steps to install Catch2v3 locally

- `git clone https://github.com/catchorg/Catch2.git -b v3.8.1 --depth 1`
- `cd Catch2`
- `mkdir build && cd build`
- `cmake .. -DCMAKE_BUILD_TYPE=Release`
- `make -j8`

Install step, you might need to have superuser permission to be able to install globally.

> The default install location for linux is `/usr/local/`.

Reviewer: ```suggestion
During installation, you may need superuser privileges to perform a global install. On Linux systems, the default installation path is `/usr/local/`.
```

- `make install`

With Catch2 installed globally we can build individual Catch2 test like this:

```bash
export ROCM_SYSTEMS_DIR=<path_to_rocm_systems>
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/external/picojson                    \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  --offload-arch=native                                                              \
  -L /usr/local/lib -lCatch2 -lCatch2Main                                            \
  -x hip <path_to_test>
```

> The command above builds hip without the `main` from hip-tests, if the purpose is just to run the test with as simple commands as possible, go with this.

Reviewer: ```suggestion
The command above builds HIP without the `main` provided by hip-tests. If your goal is simply to run the tests using the most straightforward commands, this is the option to go with.
```

If you want to use `main` provided by hip-tests, use the following command:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/external/picojson                    \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/main.cc              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/hip_test_context.cc  \
  -L /usr/local/lib                                                                  \
  --offload-arch=native                                                              \
  -lCatch2                                                                           \
  -x hip <path_to_test>
```

### Steps to use FetchContent to get Catch2

Build Catch2 with hip-tests

```bash
export HIP_PATH=<path_where_hip_is_installed>
cd rocm-systems
export ROCM_SYSTEMS_DIR=$PWD
mkdir tests
cd tests
cmake ../projects/hip-tests/catch -DHIP_PLATFORM=amd -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=$HIP_PATH
make -j$(nproc) Catch2
export DEPS_PATH=$PWD/_deps
```

> We now use the Catch2 we just build to link to the tests.

Reviewer: ```suggestion
We now use the Catch2 library we just built to link against the tests.
```

  $DEPS_PATH/catch2-src/src/catch2/internal/catch_main.cpp                           \
  -lCatch2                                                                           \
  -x hip <test_file.cc>
```

To use our main:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/external/picojson                    \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  -I $DEPS_PATH/catch2-src/src                                                       \
  -I $DEPS_PATH/catch2-build/generated-includes                                      \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/main.cc              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/hip_test_context.cc  \
  --offload-arch=native                                                              \
  -L $DEPS_PATH/catch2-build/src                                                     \
  -lCatch2                                                                           \
  -x hip <test_file.cc>
```

> You might need to set `LD_LIBRARY_PATH` to be Catch2 location.

Reviewer: ```suggestion
You may need to set `LD_LIBRARY_PATH` to point to Catch2 location.
```

* rocm x.yy: These are the stable releases based on the ROCM release.
  This type of release is typically made once a month.

## Build HIP Catch tests

For building HIP from source, please check instructions on the [HIP page](https://rocm.docs.amd.com/projects/HIP/en/latest/install/build.html).

HIP catch tests can be built via the following instructions:

1. Clone the hip-tests source code from the repository, with definition of branch. The default branch is `develop`, as an example,
```bash
$ git clone -b develop https://github.com/ROCm/rocm-systems.git
$ cd rocm-systems/projects/hip-tests
$ export HIP_TESTS_DIR="$(readlink -f .)"
```
`hip-tests` for AMD platform now rely on `amdclang` to build, which is shipped with ROCm installation.
Although individual tests will compile with `hipcc`, ideally you should use `amdclang++`.

2. Build the catch tests
```bash
$ cd "$HIP_TESTS_DIR"
$ mkdir -p build; cd build
$ cmake ../catch -DHIP_PLATFORM=amd -DCMAKE_PREFIX_PATH=<HIP-Installed-Path> -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER=amdclang++ -DCMAKE_C_COMPILER=amdclang -DCMAKE_HIP_COMPILER=amdclang++ -DOFFLOAD_ARCH_STR="--offload-arch=gfxXXXX"

> $ make -j32 build_tests

Reviewer 2: Use `-j$(nproc)` for consistency.

$ ctest # run tests
```

> 

Reviewer: Need to put note after the build command:
Note: Developers can run `rocminfo` to determine the gfxXXXX architecture from the agent name. For example:
```bash
$ /opt/rocm/bin/rocminfo
...
*******                  
Agent 2                  
*******                  
  Name:                    gfx1200
```

Besides, if the HIP compiler cannot build even a simple test program, developers may need to explicitly set the paths for `amdclang` and `amdclang++` to point to `/opt/rocm/bin`.

Reviewer 2: > Besides, if the HIP compiler cannot build even a simple test program, developers may need to explicitly set the paths for amdclang and amdclang++ to point to /opt/rocm/bin.

Could you clarify? I thought the HIP compiler can build a simple test, and `amdclang`/`amdclang++` come with ROCm (same path as `hipcc`).

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR50

HIP catch tests are built under the folder `$HIP_TESTS_DIR/build`.

### Build HIP Catch2 standalone test

`standalone_main.cc`, which was previously used, has been removed. We now rely on the main function provided by Catch2 to build standalone tests. The test suite has been upgraded to Catch2 v3 (specifically v3.8.1). This transition from v2 to v3 introduced several significant changes in how Catch2 integrates with hip-tests. Most importantly, Catch2 is no longer a single-header library; it is now a compiled library that must be linked into the test executable.

#### Getting Catch2

##### Option 1: System-wide install (recommended)

For developers working on personal machines, installing Catch2 system-wide is strongly recommended. This avoids the download and build steps within hip-tests and leads to noticeably faster build times.

- `git clone https://github.com/catchorg/Catch2.git -b v3.8.1 --depth 1`
- `cd Catch2`
- `mkdir build && cd build`
- `cmake .. -DCMAKE_BUILD_TYPE=Release`
- `make -j8`

> During installation, you may need superuser privileges to perform a global install. On Linux systems, the default installation path is `/usr/local/`.

Reviewer: This sentence should appear immediately after the `sudo make install` command, formatted as follows:

Note: On Linux systems, the default installation path is /usr/local/.

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR83

- `make install`

##### Option 2: FetchContent (build from source)

If you prefer not to install Catch2 system-wide, you can build it via cmake's FetchContent mechanism. The hip-tests cmake will automatically download and build Catch2 if it is not found on the system. To build just the Catch2 library:

```bash

> export HIP_PATH=<path_where_hip_is_installed>

Reviewer 2: Inconsistent with others; they use `HIP_DIR`.

Reviewer 2: HIP_DIR is used for the source code path.

cd rocm-systems
export ROCM_SYSTEMS_DIR=$PWD
mkdir tests
cd tests

> cmake ../projects/hip-tests/catch -DHIP_PLATFORM=amd -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=$HIP_PATH

Reviewer: I met build errors using this command like below:
CMake Error at CMakeLists.txt:285 (string):
  string sub-command REPLACE requires at least four arguments.
Please double check on this.

Reviewer 2: Can you try this:
```sh
$ export ROCM_SYSTEMS_DIR=$PWD
$ export HIP_PATH=/opt/rocm
$ mkdir tests
$ cd tests
$ hipconfig --version
$ cmake ../projects/hip-tests/catch -DHIP_PLATFORM=amd -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=$HIP_PATH
```

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR96

> make -j$(nproc) Catch2

Reviewer 2: It looks like, `make -j Catch2WithMain` is missing here.

export DEPS_PATH=$PWD/_deps
```

The Catch2 source and build artifacts will be under `$DEPS_PATH`. You may need to set `LD_LIBRARY_PATH` to point to the Catch2 library location.

#### Standalone build commands

The examples below assume `ROCM_SYSTEMS_DIR` is set to the root of the rocm-systems checkout:

```bash
export ROCM_SYSTEMS_DIR=<path_to_rocm_systems>
```

##### Without hip-tests main

The simplest way to compile a single test. This uses Catch2's own `main` and does not include the hip-tests test context, parameter store, or event listener.

With a system-wide Catch2 install:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  --offload-arch=native                                                              \
  -L /usr/local/lib -lCatch2 -lCatch2Main                                            \
  -x hip <path_to_test>
```

> 

Reviewer: I couldn't run this command.
besides, <path_to_test> is not clear here, and got error: error: error reading './tests': Is a directory

Reviewer 2: <path_to_test> is a test case file, eg. `projects/hip-tests/catch/unit/gl_interop/hipGLGetDevices.cc`.  Looks like this doc misses `Catch2Main`.

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR142

With FetchContent Catch2:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  -I $DEPS_PATH/catch2-src/src                                                       \
  -I $DEPS_PATH/catch2-build/generated-includes                                      \
  --offload-arch=native                                                              \
  -L $DEPS_PATH/catch2-build/src                                                     \
  $DEPS_PATH/catch2-src/src/catch2/internal/catch_main.cpp                           \
  -lCatch2                                                                           \
  -x hip <path_to_test>
```

> 

Reviewer: please double check this command. especially what needs to put as <path_to_test>

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR142

##### With hip-tests main

This includes the hip-tests test context (`hip_test_context.cc`) and `main.cc`, which provide the `TestContext` (platform/device info) and picojson-based config parsing.

With a system-wide Catch2 install:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/external/picojson                    \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/main.cc              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/hip_test_context.cc  \
  -L /usr/local/lib                                                                  \
  --offload-arch=native                                                              \
  -lCatch2                                                                           \
  -x hip <path_to_test>
```

> 

Reviewer: please double check this command. especially what needs to put as <path_to_test>

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR142

With FetchContent Catch2:

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/external/picojson                    \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  -I $DEPS_PATH/catch2-src/src                                                       \
  -I $DEPS_PATH/catch2-build/generated-includes                                      \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/main.cc              \
  -x hip $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/hipTestMain/hip_test_context.cc  \
  --offload-arch=native                                                              \
  -L $DEPS_PATH/catch2-build/src                                                     \
  -lCatch2                                                                           \
  -x hip <path_to_test>
```

> 

Reviewer: please double check this command. especially what needs to put as <path_to_test>

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/changes/004e7950c39f5bab0ecaa5f58555dd9dcd3b86ef#diff-adde73363fcf0710ccf0551f3670dfe1cd8c9422eb39cb930e6700d6df659f6dR142

> ©2026 Advanced Micro Devices, Inc. All Rights Reserved.

Reviewer 2: https://github.com/ROCm/rocm-systems/pull/2198/commits/2a9b45c4aa61f72e8a5b6ba70fd919d7a1f0099e

```bash
cd rocm-systems
export ROCM_SYSTEMS_DIR=$PWD
export HIP_PATH=/opt/rocm
mkdir -p tests
cd tests
cmake ../projects/hip-tests/catch -DHIP_PLATFORM=amd -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=$HIP_PATH
make -j$(nproc) Catch2WithMain
export DEPS_PATH=$PWD/_deps
```

Outputs are under `$DEPS_PATH`; set `LD_LIBRARY_PATH` if the linker cannot find Catch2 at run time.

#### Standalone build commands

The examples below assume `ROCM_SYSTEMS_DIR` is set to the root of the rocm-systems checkout:

```bash
export ROCM_SYSTEMS_DIR=<path_to_rocm_systems>
```

##### Without hip-tests main

Compile one test `.cc` with Catch2 supplying `main`. The hip-tests test runner and its shared context are not used.

**System-wide Catch2** — link `Catch2` and `Catch2Main` (Catch2’s `main`):

```bash
amdclang++                                                                           \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  --offload-arch=native                                                              \

>   -L /usr/local/lib -lCatch2 -lCatch2Main                                            \

Reviewer 2: This feedback appears to be invalid and based solely on this file’s content. `Catch2WithMain` produces `Catch2.so` and `Catch2Main.so`.

To disable it in a cmake build:

```bash
cmake ../catch -DENABLE_YAML_TAGS=OFF ...
```

This skips generation of `hip_tests_config.hh` and the YAML validation step, which can be useful for faster iteration or when the YAML configs are not needed.

For standalone `amdclang++` builds, `ENABLE_YAML_TAGS` is not defined by default since there is no cmake involved. To opt in, pass `-DENABLE_YAML_TAGS` on the compiler command line and provide the include path to the build directory containing the generated `hip_tests_config.hh`:

```bash
amdclang++                                                                           \
  -DENABLE_YAML_TAGS                                                                 \
  -I <build_dir>                                                                     \
  -I $ROCM_SYSTEMS_DIR/projects/hip-tests/catch/include                              \
  --offload-arch=native                                                              \

>   -L /usr/local/lib -lCatch2 -lCatch2Main                                            \

Reviewer 2: These standalone build instructions reference linking against -lCatch2Main, but hip-tests' CMake links against Catch2::Catch2WithMain for Catch2 v3. Please update the documented library name(s) to match Catch2 v3 (e.g., Catch2WithMain), otherwise the command is likely to fail on systems where Catch2Main is not produced/installed.
